x/subscription/keeper/epoch_start.go

```diff
@@ -1,22 +1,61 @@
 package keeper
 
 import (
+	"time"
+
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/lavanet/lava/x/subscription/types"
 )
 
 // EpochStart runs the functions that are supposed to run in epoch start
 func (k Keeper) EpochStart(ctx sdk.Context) {
 	// On epoch start we need to iterate through all the subscriptions and
-	// check if they just expired. Those that expire get deleted.
+	// check those whose current month just expired:
+	//
+	// - record the actual epoch of expiry (for rewards validation)
+	// - save the month's remaining CU in prev (for rewards validation)
+	// - reset the month's remaining CU to the plan's allowance
+	// - reduce remaining duration, and delete if it reaches zero
+	//
+	// Note that actual deletion is deferred by PAYMENT_GRACE_TIME_HOURS,
+	// to allow for payments for the last month of the subscription.
+	// (During this period, remaining CU of the new month is zero).
 
 	date := ctx.BlockTime().UTC()
 	subExpired := k.GetCondSubscription(ctx, func(sub types.Subscription) bool {
-		return sub.IsExpired(date)
+		return sub.IsMonthExpired(date)
 	})
 
-	// First collect, then remove (because removing may affect the iterator?)
 	for _, sub := range subExpired {
-		k.RemoveSubscription(ctx, sub.Consumer)
+		sub.PrevExpiryBlock = uint64(ctx.BlockHeight())
+		sub.PrevCuLeft = sub.MonthCuLeft
+
+		// subscription has been dead for a month: delete it
+		if sub.DurationLeft == 0 {
+			k.RemoveSubscription(ctx, sub.Consumer)
+			continue
+		}
+
+		sub.DurationLeft -= 1
+
+		if sub.DurationLeft > 0 {
+			date = nextMonth(date)
+			sub.MonthExpiryTime = uint64(date.Unix())
+			sub.MonthCuLeft = sub.MonthCuTotal
```

Comment:
i feel like fixation will be a better fit and will solve all of this. wdyt?
how do you know where to add the payment if they are received from the last month?

Reply:
but how do we know which is which? if a provider gives a relay payment with blockheight X, how do you know to which month to add it?

Reply:
If the payment is for this current month, then we use MonthCuLeft.
If the payment is for previous month, then we use PrevCuLeft.
We can tell which case it is because we know PrevExpiryBlock - the barrier between previous and current month.